Refactor vault

toydarian opened the pull request:
##### SUMMARY
This is a refactoring-only PR for vault.

I plan to add some quality-of-life improvements and features to vault, but before doing so, I wanted to split up the code from one 1.2k-lines file into several files and make it a little more modern.

Those are the major changes in this PR:

 - all classes and most functions have been grouped by use
 - old '%'-style format-strings have been updated to "fancy" formatted string-literals
 - fixed formatting and typos in doc-strings, comments and messages

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

or more like
- Refactoring Pull Request

##### COMPONENT NAME
vault

##### ADDITIONAL INFORMATION
Most classes and functions are not available on the package `ansible.parsing.vault` anymore, but some that are used in other components are imported in `__init__.py` to avoid making this PR even larger than it already is.

toydarian committed:
splits vault-code into multiple smaller files

s-hertel commented on lib/ansible/parsing/vault/util.py:
```diff
@@ -0,0 +1,94 @@
+    try:
+        return unhexlify(b_data)
+    except (BinasciiError, TypeError) as exc:
+        raise AnsibleVaultFormatError(f'Vault format un-hexlify error: {exc}')
```
Looks like this is causing tests to fail:

```suggestion
        raise AnsibleVaultFormatError(f'Vault format unhexlify error: {exc}')
```

toydarian replied:
Weird that this would make tests fail on some platforms, but not others.  
Thanks fir the hint, though!

s-hertel replied:
I think it failed just for all the platforms running the `ansible-vault` tests because it greps for this particular error.

s-hertel commented on lib/ansible/parsing/vault/file.py:
```diff
@@ -0,0 +1,164 @@
+    def __repr__(self):
+        if self.filename:
+            return "%s(filename='%s')" % (self.__class__.__name__, self.filename)
+        return "%s()" % self.__class__.__name__
```
Looks like these format strings were not updated.

toydarian replied:
fixed in [ae8bc60](https://github.com/ansible/ansible/pull/80675/commits/ae8bc6044a5afa7b0783dbab39f40d6d139527d3)

s-hertel commented on lib/ansible/parsing/vault/editor.py:
```diff
@@ -0,0 +1,413 @@
+        except Exception as e:
+            # if an error happens, destroy the decrypted file
+            self._shred_file(tmp_path)
+            raise AnsibleError(f'Unable to execute the command "{" ".join(cmd)}": {to_native(e)}')
```
`to_native` is unnecessary now with the f string. You could raise the AnsibleError from the original exception:
```suggestion
            raise AnsibleError(f'Unable to execute the command "{" ".join(cmd)}": {e}') from e
```

toydarian replied:
fixed in [4a27827](https://github.com/ansible/ansible/pull/80675/commits/4a27827f66f47692ecbff9058aedf38cd6d57d8a)

s-hertel commented on lib/ansible/parsing/vault/util.py:
```diff
@@ -0,0 +1,94 @@
+
+
+class AnsibleVaultFormatError(AnsibleError):
+    pass
```
These could be moved to `lib/ansible/errors/__init__.py` probably.

toydarian replied:
fixed in [ab6f9d3](https://github.com/ansible/ansible/pull/80675/commits/ab6f9d31d95d0295990316205ffa4d79aa177eca)

s-hertel reviewed (commented):
I've done an initial review and it looks like all changes are cosmetic and generally looks good to me :+1: This will need a thorough review every time it needs to be rebased, so if you can remove unnecessary changes that aren't an explicit goal of this PR (for example, modified function signatures/variable names/whitespace), that would make reviewing code parity easier.

It looks like a bunch of class methods have been turned into staticmethods, but none of them are called without a class instance. Until that's needed for some reason, those changes can be removed.

Since `lib/ansible/parsing/vault/__init__.py` is internal API, the backwards incompatible changes to the classes/methods available may not matter, but I'd like to mention it was easy to find some things in the wild that would break from that. Some examples: https://engineering.adjust.com/post/secrets_management_with_ansible_vault/#vault-python-api, https://github.com/nuagenetworks/nuage-metroae/blob/master/encrypt_credentials.py#L8, https://github.com/arenadata/adcm/blob/develop/python/cm/adcm_config/ansible.py#L14

toydarian commented:
> I've done an initial review and it looks like all changes are cosmetic and generally looks good to me +1 This will need a thorough review every time it needs to be rebased, so if you can remove unnecessary changes that aren't an explicit goal of this PR (for example, modified function signatures/variable names/whitespace), that would make reviewing code parity easier.

The few changes to variable names, signatures etc. are meant to make the code easier to understand. I already tried to keep those to a minimum.

> It looks like a bunch of class methods have been turned into staticmethods, but none of them are called without a class instance. Until that's needed for some reason, those changes can be removed.

I changed a few functions that didn't use `self` into static methods (`@staticmethod`), but I don't think I changed any `@classmethod` to `@staticmethod`.

> Since `lib/ansible/parsing/vault/__init__.py` is internal API, the backwards incompatible changes to the classes/methods available may not matter, but I'd like to mention it was easy to find some things in the wild that would break from that. Some examples: https://engineering.adjust.com/post/secrets_management_with_ansible_vault/#vault-python-api, https://github.com/nuagenetworks/nuage-metroae/blob/master/encrypt_credentials.py#L8, https://github.com/arenadata/adcm/blob/develop/python/cm/adcm_config/ansible.py#L14

Do you suggest, I import every class and metod in `__init__.py` to preserve backwards-compatibility?

s-hertel commented:
Thanks for the quick update.

> The few changes to variable names, signatures etc. are meant to make the code easier to understand. I already tried to keep those to a minimum.

I didn't want to make a comprehensive list of these because they're all very minor, but there are quite a few. To keep this general:
- The class methods (not to be confused with `classmethod`) that were decorated with `staticmethod`  fall into this category. They're still being called via class instances, so the change is unnecessary and I don't think it's an improvement to readability. I'd also question whether some of them would be better as standalone functions, if they needed to be called without a class instance.
- `vaulttext` was inconsistently renamed to `vault_text` in variable and function names, but it wasn't changed everywhere and none of the occurrences of `ciphertext` or `plaintext` were modified. Variables named consistently makes it easier to find/remember them, so I don't love this change even though I can see the argument that the double t is less readable without the underscore separating them.
- Some otherwise unchanged lines add or remove line breaks.

Ideally to me, most of these would be removed so it's easier to see that yes, this function was just moved from file1 to file2 with no actual changes other than format strings.

> Do you suggest, I import every class and metod in __init__.py to preserve backwards-compatibility?

I'm not sure, maybe. You could wait for others to weigh in on this.

toydarian commented:
> The class methods (not to be confused with `classmethod`) that were decorated with `staticmethod`  fall into this category. They're still being called via class instances, so the change is unnecessary and I don't think it's an improvement to readability. I'd also question whether some of them would be better as standalone functions, if they needed to be called without a class instance.

I have a strong opinion about this. Methods, that don't use `self` should be either static or standalone. When I call a method with `object.method()` I expect it to depend on the object.  
I changed the methods that I made static earlier to standalone functions and all usages of those. They were only used in the `vault` module or in tests, so I think that makes the most sense.

> `vaulttext` was inconsistently renamed to `vault_text` in variable and function names, but it wasn't changed everywhere and none of the occurrences of `ciphertext` or `plaintext` were modified. Variables named consistently makes it easier to find/remember them, so I don't love this change even though I can see the argument that the double t is less readable without the underscore separating them.

I'm in slight favor of separating words for readability, but I guess there are enough changes in this PR, so I reverted all `vault_text` to `vaulttext`.

> Some otherwise unchanged lines add or remove line breaks.

Those should mostly be docstrings and comments. The starting quotes were inconsistent, sometimes on a separate line, and sometimes not. That itched me a little and I find it more optically pleasing if they have their own lines. As I like to have those things consistent, I formatted them all the same way.  
I also dislike the first line of a class or function being empty and if there is an empty line after each line of code.

> Do you suggest, I import every class and method in `__init__.py` to preserve backwards-compatibility?
> 
> I'm not sure, maybe. You could wait for others to weigh in on this.

What is the best way forward in this case? Should I post to a mailing-list?

s-hertel commented on lib/ansible/parsing/vault/editor.py:
```diff
@@ -293,121 +208,205 @@ def rekey_file(self, filename, new_vault_secret, new_vault_id=None):
-            if filename == '-':
-                data = sys.stdin.buffer.read()
+
+def _shred_file_custom(tmp_path):
```
I may have been unclear, but since one of the main concerns about this PR is going to be unintended changes, the unnecessary refactoring which isn't 1) relocating code 2) updating format strings 3) improving readability, just make it harder to review the PR. These were fine in their initial state.

s-hertel replied:
I hadn't seen your comment right above, but the issue was not whether or not a staticmethod would have been better, it's not to introduce unnecessary changes that make it harder to review.

> What is the best way forward in this case? Should I post to a mailing-list?

Probably just need consensus from the core team, I don't have a strong opinion but would personally err on the side of caution. But people shouldn't be using it. Probably just need some time for others to review.

toydarian replied:
Okay, so your point is, that I should split this PR into multiple smaller ones, so they are easier to review?

bcoca replied:
it will definitely cut down the time for review/merge, i've already stopped mid way reviewing this PR 3 times because something else had to be addressed

toydarian replied:
done

s-hertel replied:
It looks like this is still here and a few others:
- _editor_shell_command
- _real_path
- _shred_file
- _shred_file_custom
- confirm
- read_data
- shuffle_files
- write_data

I excluded these from my review. I definitely think we should give code-rewriting the focus in targeted PRs instead of lumping more meaningful changes into this large refactor.

The blanket approach of first just making them all staticmethods, and then standalone functions doesn't feel well-thought out, so if you work on this as a follow-up, I think they should be considered on an individual basis, and be a deliberate code reorganization than just trying to delete the unused `self` any way possible. On the devel branch's version of this function there's a TODO to move it into its own class instead of this, so we should take into account the original intent too.

s-hertel commented on test/units/parsing/vault/test_vault.py:
```diff
@@ -19,6 +19,7 @@
 
 # Make coding more python3-ish
 from __future__ import (absolute_import, division, print_function)
+
```
```suggestion
```

s-hertel commented on test/sanity/ignore.txt:
```diff
@@ -112,7 +112,7 @@ lib/ansible/module_utils/six/__init__.py replace-urlopen
 lib/ansible/module_utils/urls.py pylint:disallowed-name
 lib/ansible/module_utils/urls.py replace-urlopen
-lib/ansible/parsing/vault/__init__.py pylint:disallowed-name
+lib/ansible/parsing/vault/editor.py pylint:disallowed-name
```
This ignore is for a dummy variable named `_`. I think this would be a reasonable one to rename so we can remove the ignore and prevent accidentally adding more.

s-hertel commented on lib/ansible/parsing/vault/lib.py:
```diff
@@ -0,0 +1,328 @@
+    try:
+        # Make sure we have a byte string and that it only contains ascii
+        # bytes.
+        b_data = to_bytes(to_text(data, encoding='ascii', errors='strict', nonstring='strict'), encoding='ascii',
+                          errors='strict')
```
```suggestion
        b_data = to_bytes(to_text(data, encoding='ascii', errors='strict', nonstring='strict'), encoding='ascii', errors='strict')
```

toydarian replied:
this would cause tests to fail, as the line is too long

s-hertel replied:
It shouldn't. My suggestion was the original line, which is < 160 characters.

toydarian replied:
you're right, my editor is set to 120 characters

s-hertel commented on lib/ansible/parsing/vault/lib.py:
```diff
@@ -0,0 +1,328 @@
+        vault_id_matchers = []
+
+        if vault_id:
+            display.vvvvv(f'Found a vault_id ({to_text(vault_id)}) in the vault-text')
```
```suggestion
            display.vvvvv(f'Found a vault_id ({to_text(vault_id)}) in the vaulttext')
```

s-hertel commented on lib/ansible/parsing/vault/lib.py:
```diff
@@ -0,0 +1,328 @@
+            this_cipher = CIPHER_MAPPING[cipher_name]()
+        else:
+            raise AnsibleError(f"{cipher_name} cipher could not be found")
+
```
Please put back initializing `b_plaintext` and `vault_id_used` in the outer scope. I think removing it makes it easier to potentially introduce issues in the try/except below.

s-hertel commented on lib/ansible/parsing/vault/file.py:
```diff
@@ -0,0 +1,168 @@
+        if popen.returncode == self.VAULT_ID_UNKNOWN_RC:
+            raise AnsibleError(
+                f'Vault password client script {self.filename} '
+                f'did not find a secret for vault-id={self._vault_id}: stderr')
```
```suggestion
                f'did not find a secret for vault-id={self._vault_id}: {stderr}')
```

s-hertel commented on lib/ansible/parsing/vault/editor.py:
```diff
@@ -0,0 +1,417 @@
+        b_vaulttext = read_data(filename)
+        vaulttext = to_text(b_vaulttext)
+
+        display.vvvvv(f'Re-keying file "{to_text(filename)}" to with new vault-id "{to_text(new_vault_id)}" '
```
```suggestion
        display.vvvvv(f'Rekeying file "{to_text(filename)}" to with new vault-id "{to_text(new_vault_id)}" '
```

s-hertel commented on lib/ansible/parsing/vault/ciphers.py:
```diff
@@ -0,0 +1,238 @@
+    except AnsibleVaultFormatError:
+        raise
+    except Exception as exc:
+        raise AnsibleVaultFormatError(f"Vault text format error: {exc}") from exc
```
```suggestion
        raise AnsibleVaultFormatError(f"Vault vaulttext format error: {exc}") from exc
```

s-hertel commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -357,943 +65,17 @@ def get_file_vault_secret(filename=None, vault_id=None, encoding=None, loader=No
+        if file.script_is_client(filename):
             # this is special script type that handles vault ids
-            display.vvvv(u'The vault password file %s is a client script.' % to_text(this_path))
-            # TODO: pass vault_id_name to script via cli
```
Please leave the TODO.

toydarian replied:
I'm pretty sure that is already done. The vault-id is passed to the script via cli.
```python
    def _build_command(self):
        command = [self.filename]
        if self._vault_id:
            command.extend(['--vault-id', self._vault_id])

        return command
```
https://github.com/ansible/ansible/pull/80675/files#diff-7469b009ae7ba9579568eae1a30e81f0dd143311139d6ccca88a3dcf49a48447R144-R149

s-hertel replied:
You're right, thanks for spotting that. The commit that added `_build_command` moved around that comment, but forgot to remove it I think.

s-hertel reviewed (commented):
Thanks for reducing some of the extraneous changes. Still LGTM in general, but there are a few more minor changes that could be tidied up.

s-hertel commented on lib/ansible/parsing/vault/secret.py:
```diff
@@ -0,0 +1,105 @@
+        return None
+
+    @staticmethod
+    def confirm(b_vault_pass_1, b_vault_pass_2):
```
```suggestion
    def confirm(self, b_vault_pass_1, b_vault_pass_2):
```

s-hertel commented on lib/ansible/parsing/vault/editor.py:
```diff
@@ -0,0 +1,411 @@
+        os.chown(filename, prev.st_uid, prev.st_gid)
+
+        display.vvvvv(
+            f'Re-keyed file "{to_text(filename)}" (decrypted with vault id "{to_text(vault_id_used)}") was encrypted '
```
```suggestion
            f'Rekeyed file "{to_text(filename)}" (decrypted with vault id "{to_text(vault_id_used)}") was encrypted '
```

s-hertel commented on lib/ansible/parsing/vault/ciphers.py:
```diff
@@ -0,0 +1,238 @@
+# (c) 2014, James Tanner <tanner.jc@gmail.com>
+# (c) 2016, Adrian Likins <alikins@redhat.com>
+# (c) 2016 Toshio Kuratomi <tkuratomi@ansible.com>
+# (c) 2023, Thomas Ziegler <thomas.ziegler.pa@gmail.com>
```
I don't think rewrites/refactoring usually warrant a copyright addition (same for the other files). Adding your name would be more applicable on a PR containing feature work.

s-hertel reviewed (approved):
The git blame with `-C` looks great now. I don't think there's a way to see from the UI, so here's the current state of things squashed to a single commit: https://gist.github.com/s-hertel/6f7038adf383a7746269da94943dcf0e.

toydarian commented:
Hello @s-hertel, do you have a suggestion on how I can get attention to this MR? I posted on the chat a few weeks back, but to no success.

nitzmahone commented:
(needs review, major changes to vault coming)

toydarian commented:
> (needs review, major changes to vault coming)

@nitzmahone is this going to get attention? I had given up on it, but if it is going to be picked up, I'm going to rebase and resolve conflicts

bcoca commented:
@toydarian no, the point is that this PR might need to be redone or even made irrelevant with upcoming changes. We have kept it open waiting for the relevant core members having some time to do a deep review and comparison to make a final decision.
